File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // using two loops
    /* for(let i=0;i<nums.length-1;i++) {
      for(let j=i+1;j<nums.length;j++) {
         if(nums[i]+nums[j]==target){
            return new Array(i,j);
         }
      }
    } */
    // using map
    const map = new Map();
     for(let i=0;i<nums.length;i++){
        let rem_value = target - nums[i]; 
        // first check whether remaining value is present in the hmap
        if(map.has(rem_value)) {
            let rem_value_index = map.get(rem_value)
            return new Array(rem_value_index,i)
        }
        else{
            map.set(nums[i], i)
        }
    }
};

File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // have to return the indices so we cannot use sorting
        int n = nums.length;
        /*
        // Brute Force Approach - using two loops
        for(int i=0;i<n-1;i++) {
            for(int j=i+1;j<n;j++) {
                int sum = nums[i]+nums[j];
                if(sum == target) {
                    return new int[]{i, j};
                }
            }
        }
        */
        // Optimized Approach - using HashMap
        HashMap<Integer,Integer> hmap = new HashMap<>();
        // insert element value and index in hashmap
        for(int i=0;i<n;i++) {
            int complement = target - nums[i];
            if(hmap.containsKey(complement)) {
                return new int[]{hmap.get(complement), i};
            }
            hmap.put(nums[i], i);
        }
        return new int[]{}; 
    }
}